### PR TITLE
Update to latest Substrate master.

### LINKED ISSUE #221: Change `BlockNumber` to u32
https://github.com/paritytech/polkadot/blob/master/primitives/src/lib.rs#L62

### PR DESCRIPTION
**Pull request by kianenigma**
Closes #221 

TODO:

- [x] Fix all build / test errors.
- [x] `FindAuthor`.
- [x] Service and runtime are still using `aura`. Both should be replaced with `babe` now.
- [x] #342.
- [x] Fix initial authorities in chain_spec (needs quite some work).
- [x] `system::CheckGenesis<Runtime>`
- [x] fix and verify that a `--dev` chain works
  - [x] compiles and imports blocks.
  - [x] submit and process extrinsic
- [x] fix and verify that a alice/bob chain works
  - [x] compiles and imports blocks.
  - [x] submit and process extrinsic

cc @jacogr I am doing the testing with UI along the way but if you also want to give it a try would be great.

### CODE REVIEW

**Comment by kianenigma** [comment]
[clabot:check]

**Comment by kianenigma** [comment]
[clabot:check]

**Comment by kianenigma** [comment]
I will stall this for now halfway through integrating BABE until https://github.com/paritytech/substrate/pull/3296 lands. There seems to a lot of assumptions made in the polkadot code about crypto concrete types which might be affected by the new key mgmt system.

**Comment by gavofyork** [comment]
@kianenigma will this include the content of #342 ?

**Comment by kianenigma** [comment]
> @kianenigma will this include the content of #342 ?

Ideally yes, It is on my todo list up top.

File: collator/src/lib.rs [hunk 1]
```diff
@@ -364,7 +363,9 @@ impl<P, E> Worker for CollationNode<P, E> where
 						None => return future::Either::A(future::ok(())),
 					};
 
-					let authorities = try_fr!(api.authorities(&id));
+					// TODO: double check this. It makes sense bc other occurances of `authorities`
+					// has been refactored to this.
+					let authorities = try_fr!(api.validators(&id));
```
Comment (kianenigma) [inline collator/src/lib.rs]: @rphmeier is this okay?

File: primitives/src/lib.rs [hunk 2]
```diff
@@ -69,12 +72,13 @@ pub type SessionKey = ed25519::Public;
 /// that 32 bits may be multiplied with a balance in 128 bits without worrying about overflow.
 pub type Balance = u128;
 
+// TODO: I don't get why this was defined here. It was aura so I change it to babe but regardless of that
```
Comment (kianenigma) [inline primitives/src/lib.rs]: yeah, I still agree with my note. _If the crypto key type is defined inside the `babe` module and exported via `babe::AuthorityId` then that should be used everywhere_. I don't get why the other one is needed.

File: runtime/Cargo.toml [hunk 3]
```diff
@@ -41,14 +39,16 @@ system = { package = "srml-system", git = "https://github.com/paritytech/substra
 timestamp = { package = "srml-timestamp", git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-master" }
 treasury = { package = "srml-treasury", git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-master" }
 version = { package = "sr-version", git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-master" }
+babe = { package = "srml-babe", git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-master" }
```
Comment (kianenigma) [inline runtime/Cargo.toml]: I hate this myself. Should refactor them to:
- `babe` crate should always point to the `core` portion
- `srml_babe` should point to the srml portion.

File: service/src/chain_spec.rs [hunk 4]
```diff
@@ -34,33 +36,46 @@ const DEFAULT_PROTOCOL_ID: &str = "dot";
...
-		hex!["42d69e4222c08885a4d6ff65f01852ba4a1599b683ad66286e4603d825e26b49"].unchecked_into(), // 5DaLmkrGTFvSTHBpShqqqRYaydw1sT4u3NCogYiE8Q1LqUUp
+		hex!["46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a"].unchecked_into(), // 5DaLmkrGTFvSTHBpShqqqRYaydw1sT4u3NCogYiE8Q1LqUUp
 	];
 
 	// for i in 1 2 3 4; do for j in stash controller; do subkey inspect "$SECRET//$i//$j"; done; done
-	// for i in 1 2 3 4; do for j in session; do subkey -e inspect "$SECRET//$i//$j"; done; done
-	let initial_authorities: Vec<(AccountId, AccountId, SessionKey)> = vec![(
+	// for i in 1 2 3 4; do for j in session; do subkey --sr25519 inspect "$SECRET//$i//$j"; done; done
+	// for i in 1 2 3 4; do for j in session; do subkey --ed25519 inspect "$SECRET//$i//$j"; done; done
+	// TODO: update these.
```
Comment (kianenigma) [inline service/src/chain_spec.rs]: yeah -- for now I just duplicated one to go forward with the compiler.

File: validation/src/lib.rs [hunk 5]
```diff
@@ -280,7 +280,9 @@ impl<C, N, P> ParachainValidation<C, N, P> where
 
 		let id = BlockId::hash(parent_hash);
 
-		let authorities = self.client.runtime_api().authorities(&id)?;
+		// TODO: This used to use AuraApi and now it uses validators from parachainHost, as
+		// recommended by Rob. Please double and triple check.
+		let authorities = self.client.runtime_api().validators(&id)?;
```
Comment (kianenigma) [inline validation/src/lib.rs]: @rphmeier any insights?

**Comment by kianenigma** [comment]
ok - everything seems to work properly. Have to see what awaits after rebasing `polkadot-master` with the latest substrate master inc. key factoring.

File: primitives/src/lib.rs [hunk 6]
```diff
@@ -69,12 +72,14 @@ pub type SessionKey = ed25519::Public;
...
+// I don't get why it was here.
 /// The aura crypto scheme defined via the keypair type.
 #[cfg(feature = "std")]
-pub type AuraPair = ed25519::Pair;
+pub type BabePair = sr25519::Pair;
```
Comment (gavofyork) [inline primitives/src/lib.rs]:
```suggestion
pub type BabePair = babe::AuthorityPair;
```

File: primitives/src/lib.rs (hunk 2 above)
Comment (gavofyork) [reply primitives/src/lib.rs]: indeed.

File: runtime/src/lib.rs [hunk 7]
```diff
@@ -191,11 +196,13 @@ parameter_types! {
 	pub const Offset: BlockNumber = 0;
 }
 
-type SessionHandlers = (Grandpa, Aura);
+type SessionHandlers = (Grandpa, Babe);
 impl_opaque_keys! {
 	pub struct SessionKeys {
 		#[id(key_types::ED25519)]
```
Comment (gavofyork) [inline runtime/src/lib.rs]:
```suggestion
		#[id(key_types::GRANDPA)]
```

File: runtime/src/lib.rs [hunk 8]
```diff
@@ -191,11 +196,13 @@ parameter_types! {
...
 		pub ed25519: GrandpaId,
+		#[id(key_types::SR25519)]
```
Comment (gavofyork) [inline runtime/src/lib.rs]:
```suggestion
		#[id(key_types::BABE)]
```

File: service/src/chain_spec.rs [hunk 9]
```diff
@@ -34,33 +37,45 @@ const DEFAULT_PROTOCOL_ID: &str = "dot";
 pub type ChainSpec = ::service::ChainSpec<GenesisConfig>;
 
 pub fn poc_3_testnet_config() -> Result<ChainSpec, String> {
-	ChainSpec::from_embedded(include_bytes!("../res/alexander.json"))
+	ChainSpec::from_json_bytes(&include_bytes!("../res/alexander.json")[..])
+}
+
+fn session_keys(ed_key: ed25519::Public, sr_key: sr25519::Public) -> SessionKeys {
```
Comment (gavofyork) [inline service/src/chain_spec.rs]: no - session keys are by application, not crypto

File: service/src/chain_spec.rs [hunk 10]
```diff
@@ -151,25 +175,33 @@ pub fn get_account_id_from_seed(seed: &str) -> AccountId {
...
+/// Helper function to generate BabeId from seed
+pub fn get_babe_id_from_seed(seed: &str) -> BabeId {
+	sr25519::Pair::from_string(&format!("//{}", seed), None)
+		.expect("static values are valid; qed")
+		.public()
+}
+
 /// Helper function to generate SessionKey from seed
-pub fn get_session_key_from_seed(seed: &str) -> SessionKey {
+pub fn get_grandpa_id_from_seed(seed: &str) -> SessionKey {
```
Comment (gavofyork) [inline service/src/chain_spec.rs]: if it is returning `SessionKey` then it's probably not `grandpa_id`

File: service/src/chain_spec.rs [hunk 11]
```diff
@@ -151,25 +175,33 @@ pub fn get_account_id_from_seed(seed: &str) -> AccountId {
...
 	ed25519::Pair::from_string(&format!("//{}", seed), None)
 		.expect("static values are valid; qed")
 		.public()
 }
 
 /// Helper function to generate stash, controller and session key from seed
-pub fn get_authority_keys_from_seed(seed: &str) -> (AccountId, AccountId, SessionKey) {
+pub fn get_authority_keys_from_seed(seed: &str) -> (AccountId, AccountId, BabeId, GrandpaId) {
```
Comment (gavofyork) [inline service/src/chain_spec.rs]: this will need to return `ParachainId`, too.

File: service/src/lib.rs [hunk 12]
```diff
@@ -150,7 +171,7 @@ impl PolkadotService for Service<LightComponents<Factory>> {
 service::construct_service_factory! {
 	struct Factory {
 		Block = Block,
-		ConsensusPair = AuraPair,
+		ConsensusPair = BabePair,
```
Comment (gavofyork) [inline service/src/lib.rs]: shouldn't be here any more.

File: service/src/lib.rs [hunk 13]
```diff
@@ -150,7 +171,7 @@ impl PolkadotService for Service<LightComponents<Factory>> {
...
 		FinalityPair = GrandpaPair,
```
Comment (gavofyork) [inline service/src/lib.rs]: shouldn't be here any more.

File: service/src/lib.rs [hunk 14]
```diff
@@ -170,19 +191,30 @@ service::construct_service_factory! {
...
+					}
+				}
+
 				// always run GRANDPA in order to sync.
 				{
 					let grandpa_key = if service.config.disable_grandpa {
 						None
 					} else {
-						service.authority_key()
+						service.fg_authority_key()
```
Comment (gavofyork) [inline service/src/lib.rs]: gaaa please no!!!

File: service/src/lib.rs [hunk 15]
```diff
@@ -231,8 +263,8 @@ service::construct_service_factory! {
 				};
 
 				// run authorship only if authority.
-				let aura_key = match service.authority_key()  {
-					Some(key) => Arc::new(key),
+				let babe_key = match service.authority_key()  {
```
Comment (gavofyork) [inline service/src/lib.rs]: pretty sure this isn't right any more.

File: service/src/lib.rs [hunk 16]
```diff
@@ -290,53 +322,60 @@ service::construct_service_factory! {
...
+					polkadot_runtime::constants::time::SLOT_DURATION,
 					service.config.custom.max_block_data_size,
 				);
 
-				info!("Using authority key {}", aura_key.public());
-				let task = start_aura(
-					SlotDuration::get_or_compute(&*client)?,
-					aura_key,
-					client.clone(),
+				info!("Using authority key for BABE: {}", babe_key.public());
```
Comment (gavofyork) [inline service/src/lib.rs]: Don't see how this can be right.

File: validation/src/lib.rs [hunk 17]
```diff
@@ -58,7 +58,7 @@ use futures03::{future::{self, Either, FutureExt}, task::Context, stream::Stream
 use collation::CollationFetch;
 use dynamic_inclusion::DynamicInclusion;
 use inherents::InherentData;
-use runtime_aura::timestamp::TimestampInherentData;
+use runtime_babe::timestamp::TimestampInherentData;
 use log::{info, debug, warn, trace, error};
 
 use ed25519::Public as AuthorityId;
```
Comment (gavofyork) [inline validation/src/lib.rs]: should be creating its own keys, like in babe and grandpa.

**Comment by gavofyork** [comment]
@kianenigma this doesn't actually update to the latest substrate. to do that you need to update `polkadot-master` branch.

**Comment by rphmeier** [comment]
in 
277be3f20dc72c340806234b340021ca674761c2

adding `sr-primitives` as a dependency to `polkadot-parachain` won't fly. it's forcing every parachain to have Substrate's runtime scaffolding libs as a dependency, which we don't want.

**Comment by kianenigma** [comment]
> @kianenigma this doesn't actually update to the latest substrate. to do that you need to update `polkadot-master` branch.

last time I checked I made sure that it has the crypto updates of substrate --just in case I updated `polkadot-master` again with the latest changes.

File: primitives/src/parachain.rs [hunk 18]
```diff
@@ -27,32 +27,53 @@ use serde::{Serialize, Deserialize};
...
+#[cfg(feature = "std")]
+pub type CollatorPair = collator_app::Pair;
 
 /// Signature on candidate's block data by a collator.
-pub type CollatorSignature = ed25519::Signature;
+pub type CollatorSignature = collator_app::Signature;
+
+pub const PARACHAIN_KEY_TYPE_ID: KeyTypeId = KeyTypeId(*b"para");
+
+mod validator_app {
```
Comment (kianenigma) [inline primitives/src/parachain.rs]: why not call it `parachain_app`?

Comment (kianenigma) [reply primitives/src/parachain.rs]: or parachain_validator (to keep the consistency of bytes given to `KeyTypeId` with the name of the app).

File: service/src/chain_spec.rs (hunk 10 above)
Comment (kianenigma) [reply service/src/chain_spec.rs]: This file _might still work_ but __for sure__ has a confused terminology now that the new crypto-app-stuff is in. Someone should have a look at it later.

File: primitives/src/parachain.rs (hunk 18 above)
Comment (rphmeier) [reply primitives/src/parachain.rs]: well, it is in the `parachain::validator_app` module.

File: collator/src/lib.rs [hunk 19]
```diff
@@ -364,18 +363,20 @@ impl<P, E> Worker for CollationNode<P, E> where
 						None => return future::Either::A(future::ok(())),
 					};
 
-					let authorities = try_fr!(api.authorities(&id));
+					// TODO: double check this. It makes sense bc other occurances of `validators`
```
Comment (shawntabrizi) [inline collator/src/lib.rs]: Did someone double check this?

Comment (gavofyork) [inline collator/src/lib.rs]: @rphmeier you ok with this?

File: runtime/src/parachains.rs [hunk 20]
```diff
@@ -528,6 +531,7 @@ impl<T: Trait> Module<T> {
 	pub fn calculate_duty_roster() -> (DutyRoster, [u8; 32]) {
 		let parachains = Self::active_parachains();
 		let parachain_count = parachains.len();
+		// TODO: use decode length.
```
Comment (gavofyork) [inline runtime/src/parachains.rs]: ??? who put this in?

File: runtime/src/parachains.rs [hunk 21]
```diff
@@ -957,9 +963,15 @@ mod tests {
 		type MinimumPeriod = MinimumPeriod;
 	}
 
-	impl aura::Trait for Test {
-		type HandleReport = aura::StakingSlasher<Test>;
-		type AuthorityId = AuraId;
+	// TODO
```
Comment (gavofyork) [inline runtime/src/parachains.rs]: TODO what?

File: collator/src/lib.rs (hunk 19 above)
Comment (rphmeier) [reply collator/src/lib.rs]: yes

Comment (rphmeier) [reply collator/src/lib.rs]: yup, it's fine

**Comment by rphmeier** [comment]
I didn't manage to get an alice/bob chain _validating_ after the key refactoring was put in. the chain spec needs a look taken at

**Comment by rphmeier** [comment]
I got alice authoring blocks. Didn't get her to connect to Bob, though.

**Pull request merged**